spinnaker/resource_application.go

```diff
@@ -151,7 +151,13 @@ func resourceSpinnakerApplicationRead(d *schema.ResourceData, meta interface{})
 }
 
 func resourceSpinnakerApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
-	clientConfig := meta.(gateConfig)
+	var clientConfig gateConfig
+	var ok bool
+	if clientConfig, ok = meta.(gateConfig); !ok {
+		// TODO(@KeisukeYamashita): This is mercari specific problem
+		// once we fixed the CI, we will remove this
+		return nil
```

Comment:
I think it's better to return error here to know the gateConfig doesn't exist.

We just want to ignore the error in initializing the gate client (providerConfigureFunc)

What do you think?

Reply:
@dragon3 That is okay when we don't run `terraform plan` (or `apply`).

But when we ran the `terraform plan` (or `apply`), if the `providerConfigureFunc` returns `nil` gate client, it will ran into panic.
Therefore, I should return `nil` not to block the `plan` or `apply` step as well.

Reply:
We definitely need nil check (`if clientConfig, ok = meta.(gateConfig); !ok {...`).

I meant it's better to return error in `resourceSpinnakerApplication*` functions.
If there is no `spinnaker_application` resource, it should not block. 
But if there is `spinnaker_application` resource, it should block.

Does it make sense?

Reply:
I meant that we should return `nil` in response to this issue.
If we return and `error` in `resourceSpinnakerApplication*` the terraform apply with fail with exit code non `0`.

The process is:

1. `terraform plan` is called by user
2.  `providerConfigureFunc` is called.
3. (When the gate client failed, ) the `nil` gate client will be passed to `resourceSpinnakerApplication*` if users has `spinnaker_application.tf` resource
4.  CRUD functions will be called. If the `nil` gate client is returned by step `3.`, the `meta.(gateConfig)` will be `nil`.
5.  if we return an `error` in `resourceSpinnakerApplication*` in the `nil` check of the `gateConfig`, the whole `plan` or `apply` will fail.

As the result, the CI will fail.
Since we want to ignore the gate client setup failure though our CI workflow, we should return `nil`, right?

Am I misunderstanding? 🙇

Reply:
In my understanding, we want:

- if there is no `spinnaker_application.tf` resource, plan and apply should NOT fail
- if there is `spinnaker_application.tf` resource, plan and apply should fail

No?

Reply:
@dragon3 In my understanding, we want: 

* Even if we Spin config is misconfigured or the Spinnaker API is down, we want to pass the `terraform plan` or `apply` (, ignore the error)

? 

> if there is no spinnaker_application.tf resource, plan and apply should NOT fail

I wrote the correct behaivour in in our internal GitHub Issue but the `providerConfigureFunc` was called at `terraform apply` or `plan` 👍

Reply:
Thank you, I understand 🙇  🙇 

Just in case for open source community: 

This is Mercari specify patch and will not be released to the main stream release.